### PR TITLE
switch xml parsing depencendy from simple-xml to xml2js

### PR DESCRIPTION
**Pull request by sebastian-kreis**

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -174,18 +174,24 @@
...
-        var body = XML.parse(res.body).Error;
-        error.code = body.Code;
-        error.message = body.Message;
-        error.requestId = body.RequestId;
-        error.hostId = body.HostId;
+      parseXMLString(res.body, function (err, result) {
+        var body;
+        if (err) {
+          debug('Error parsing error body', err.message);
+          return cb(error);
```
Comment (jtblin) [inline lib/client.js:181]: `error` is not defined here, you used `err`.

Edit: ignore that, i wasn't familiar with the code anymore.

File: lib/client.js [hunk 2]
```diff
@@ -174,18 +174,24 @@
...
+        body = result.Error;
+        error.code = body.Code[0];
+        error.message = body.Message[0];
+        error.requestId = body.RequestId[0];
+        error.hostId = body.HostId[0];
         error.key = body.Key;
-      } catch (err2) {
-        debug('Error parsing error body', err2.message);
-      } finally {
         cb(error);
```
Comment (jtblin) [inline lib/client.js]: Same here, just use `cb()`.

File: lib/client.js [hunk 3]
```diff
@@ -174,18 +174,24 @@
...
-        error.message = body.Message;
-        error.requestId = body.RequestId;
-        error.hostId = body.HostId;
+      parseXMLString(res.body, function (err, result) {
+        var body;
+        if (err) {
+          debug('Error parsing error body', err.message);
+          return cb(error);
+        }
+        if (!result) {
```
Comment (jtblin) [inline lib/client.js]: Put a space after `!` to be consistent with the rest of the code please.

File: lib/multipart.js [hunk 4]
```diff
@@ -39,15 +39,20 @@
...
       self.post(format('%s?uploads', filename), '', headers, options, function (err, res) {
         if (err) return ws.emit('error', err);
-        try {
-          var body = XML.parse(res.body).InitiateMultipartUploadResult;
+        parseXMLString(res.body, function (err, result) {
+          var body;
+          if (err) {
+            return ws.emit('error', err);
+          }
+          if (!result) {
```
Comment (jtblin) [inline lib/multipart.js]: Put a space after `!` and remove brackets to be consistent with the rest of the code please (same above for `if (err)`.

File: lib/multipart.js [hunk 5]
```diff
@@ -39,15 +39,20 @@
...
+            return ws.emit('error', new Error('Could not parse XML string'));
+          }
+          body = result.InitiateMultipartUploadResult;
           if (! body || ! body.UploadId) return ws.emit('error', new Error('Invalid upload id'));
-          ws.uploadId = body.UploadId;
+          ws.uploadId = body.UploadId[0];
```
Comment (jtblin) [inline lib/multipart.js:48]: OOC, why is that different here?

**Comment by jtblin** [comment]
Thanks. Can you please take care of the minor comments I've entered and squash your commits into one? Then I'll merge.

File: lib/multipart.js (hunk 5 above)
Comment (sebastian-kreis) [inline lib/multipart.js:48]: Hi, I am not sure what to change here.
xml2js parses the xml string:
`<InitiateMultipartUploadResult><UploadId>foobar</UploadId></InitiateMultipartUploadResult>`
into:
`{ InitiateMultipartUploadResult: { UploadId: [ 'foobar' ] } }`

Would the following change be more consistent:

```
if (! body || ! body.UploadId[0]) return ws.emit('error', new Error('Invalid upload id'));
ws.uploadId = body.UploadId[0];
```

Comment (jtblin) [inline lib/multipart.js:48]: Yes that'd be better, `! body.UploadId` would be false with an empty array. `! body.UploadId.length` would also work.

Thank you.

**Comment by sebastian-kreis** [comment]
Hi, I took care of the lines you mentioned and squashed the commits into a single one. Do I need to make a new Pull request or can you merge it like this?

**Comment by jtblin** [comment]
Like this is fine, thanks.

**Comment by jtblin** [comment]
Published `0.1.3`, thanks for the PR :+1:

**Pull request merged**